### PR TITLE
🐛clusterctl: move ensure all the namespaces exists

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR ensures all the required namespaces are in place before starting the move process

**Which issue(s) this PR fixes**
rif #1729

/area clusterctl
/milestone v0.3.0
/assign @ncdc
/assign @vincepri

### CODE REVIEW

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 1]
```diff
@@ -229,6 +236,67 @@ func setClusterPause(proxy Proxy, clusters []*node, value bool, log logr.Logger)
...
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Namespace",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		o.log.V(1).Info("Creating", ns.Kind, ns.Name)
+		if err = cs.Create(ctx, ns); err != nil && !apierrors.IsAlreadyExists(err) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
		if err := cs.Create(ctx, ns); err != nil && !apierrors.IsAlreadyExists(err) {
```

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 2]
```diff
@@ -229,6 +236,67 @@ func setClusterPause(proxy Proxy, clusters []*node, value bool, log logr.Logger)
...
+		}
+		namespaces.Insert(namespace)
+
+		// Otherwise check if namespace exists (also dealing with RBAC restrictions).
+		ns := &corev1.Namespace{}
+		key := client.ObjectKey{
+			Name: namespace,
+		}
+
+		if err = cs.Get(ctx, key, ns); err == nil {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
		if err := cs.Get(ctx, key, ns); err == nil {
```

**Comment by fabriziopandini** [comment]
@vincepri comment addressed!

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 3]
```diff
@@ -229,6 +236,67 @@ func setClusterPause(proxy Proxy, clusters []*node, value bool, log logr.Logger)
...
+			if err := cs.List(ctx, namespaces); err != nil {
+				return err
+			}
+
+			for _, ns := range namespaces.Items {
+				if ns.Name == namespace {
+					return nil
+				}
+			}
+		}
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]: Is there a particular use case that we're dealing with here? I'm wondering why this fallback to list is necessary. 

Also shouldn't we check the `Continue` ID?

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: This is making the code tolerant to use cases where the user is allowed to list namespaces but not to get namespaces (copied from the old clusterctl, and it makes sense IMO)

> Also shouldn't we check the Continue ID?
Fixed

this is not fully clear to me...

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 4]
```diff
@@ -229,6 +236,72 @@ func setClusterPause(proxy Proxy, clusters []*node, value bool, log logr.Logger)
...
+			namespaceExists := false
+			for _, ns := range namespaces.Items {
+				if ns.Name == namespace {
+					namespaceExists = true
+					break
+				}
+			}
+			if namespaceExists {
+				continue
+			}
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/mover.go]: What I meant was to check `NamespaceList` Continue flag and use `client.Continue` in the List call

Comment (vincepri) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_controller_noderef.go#L95-L120

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/mover.go]: ok, now I got it

**Comment by fabriziopandini** [comment]
@vincepri PTAL

**Comment by vincepri** [comment]
Should we add tests in a follow up issue?

**Pull request merged**